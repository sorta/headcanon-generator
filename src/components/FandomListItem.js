import React, { Component } from 'react';
import PropTypes from 'prop-types';

class FandomList extends Component {
  static propTypes = {
    fandom: PropTypes.shape({ name: PropTypes.string }),
    selectFandom: PropTypes.func,
    updateFandom: PropTypes.func,
    deleteFandom: PropTypes.func,
    isManaging: PropTypes.func,
    selectedFandomKey: PropTypes.string,
    index: PropTypes.string,
  };
  static defaultProps = {
    fandom: { name: '' },
    selectFandom: () => {},
    updateFandom: () => {},
    deleteFandom: () => {},
    isManaging: () => false,
    selectedFandomKey: '',
    index: '',
  };

  handleChange = (event) => {
    const updatedFandom = {
      ...this.props.fandom,
      [event.currentTarget.name]: event.currentTarget.value,
    };

    this.props.updateFandom(this.props.index, updatedFandom);
  };

  render() {
    const isSelected = this.props.index === this.props.selectedFandomKey;
    let itemContent = (
      <div className="control-row">
        <span className="control-stretch">{this.props.fandom.name}</span>
        <button onClick={() => this.props.selectFandom(isSelected ? '' : this.props.index)}>
          {isSelected ? '<<' : '>>'}
        </button>
      </div>
    );

    if (this.props.isManaging()) {
      itemContent = (
        <div className="control-row">
          <input name="name" type="text" placeholder="Fandom Name" className="control-stretch"
            value={this.props.fandom.name} onChange={this.handleChange}
          />
          <button onClick={() => this.props.deleteFandom(this.props.index)}>-</button>
          <button onClick={() => this.props.selectFandom(isSelected ? '' : this.props.index)}>
            {isSelected ? '<<' : '>>'}
          </button>
        </div>
      );
    }

    return (
      <li key={this.props.index} className="FandomListItem list-item">
        {itemContent}
      </li>
    );
  }
}

export default FandomList;
