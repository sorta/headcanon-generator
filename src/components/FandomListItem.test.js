import React from 'react';
import { shallow, mount } from 'enzyme';
import renderer from 'react-test-renderer';

import FandomListItem from './FandomListItem';

const props = {
  fandom: { name: 'Test 1' },
  fandomAvailability: {},
  updateFandom: jest.fn(),
  deleteFandom: jest.fn(),
  setAvailability: jest.fn(),
  isManaging: () => false,
  index: 'f_test_1',
  selectFandom: jest.fn(),
  selectedFandomKey: '',
};

// Tests for regular mode
describe("FandomListItem", () => {

  it('renders and matches our snapshot', () => {
    const snapshot = renderer.create(
      <FandomListItem { ...props } />
    );
    const tree = snapshot.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('has a FandomListItem li element', () => {
    const component = shallow(<FandomListItem { ...props } />);
    expect(component.find('li.FandomListItem.list-item').exists()).toBe(true);
  });

  it('has a checkbox and label', () => {
    const component = shallow(<FandomListItem { ...props } />);
    expect(component.find(`input[type="checkbox"][name="isAvailable"]#isAvailable_fand_${props.index}`).exists()).toBe(true);
    expect(component.find(`label[htmlFor="isAvailable_fand_${props.index}"]`).exists()).toBe(true);
  });

  it('clicking checkbox calls setAvailability', () => {
    const component = mount(<FandomListItem { ...props } />);
    component.find('input[type="checkbox"][name="isAvailable"]').simulate('change', { target: { checked: false } });
    expect(props.setAvailability.mock.calls.length).toEqual(1);
  });

  it('does not have edit elements', () => {
    const component = shallow(<FandomListItem { ...props } />);
    expect(component.find('input[name="name"]').exists()).toBe(false);
    expect(component.find('button.FandomListItem-delete').exists()).toBe(false);
  });

  it('clicking select button calls selectFandom', () => {
    const component = mount(<FandomListItem { ...props } />);
    component.find('button.FandomListItem-select').simulate('click');
    expect(props.selectFandom.mock.calls.length).toEqual(1);
  });
});

// Tests for managing mode
const mProps = { ...props };
mProps.isManaging = () => true;
describe("FandomListItem in managing mode", () => {

  it('renders and matches our snapshot', () => {
    const snapshot = renderer.create(
      <FandomListItem { ...mProps } />
    );
    const tree = snapshot.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('has a FandomListItem li element', () => {
    const component = shallow(<FandomListItem { ...mProps } />);
    expect(component.find('li.FandomListItem.list-item').exists()).toBe(true);
  });

  it('has a checkbox and label', () => {
    const component = shallow(<FandomListItem { ...mProps } />);
    expect(component.find(`input[type="checkbox"][name="isAvailable"]#isAvailable_fand_${mProps.index}`).exists()).toBe(true);
    expect(component.find(`label[htmlFor="isAvailable_fand_${mProps.index}"]`).exists()).toBe(true);
  });

  it('clicking checkbox calls setAvailability', () => {
    const component = mount(<FandomListItem { ...mProps } />);
    component.find('input[type="checkbox"][name="isAvailable"]').simulate('change', { target: { checked: false } });
    expect(mProps.setAvailability.mock.calls.length).toEqual(2);
  });

  it('has edit elements', () => {
    const component = shallow(<FandomListItem { ...mProps } />);
    expect(component.find('input[name="name"]').exists()).toBe(true);
    expect(component.find('button.FandomListItem-delete').exists()).toBe(true);
  });

  it('clicking delete button calls deleteFandom', () => {
    const component = shallow(<FandomListItem { ...mProps } />);
    component.find('button.FandomListItem-delete').simulate('click');
    expect(mProps.deleteFandom.mock.calls.length).toEqual(1);
  });

  it('changing name value calls updateFandom', () => {
    const component = mount(<FandomListItem { ...mProps } />);
    component.find('input[name="name"]').simulate('change', { target: { value: 'TEST' } });
    expect(mProps.updateFandom.mock.calls.length).toEqual(1);
  });

  it('clicking select button calls selectFandom', () => {
    const component = mount(<FandomListItem { ...mProps } />);
    component.find('button.FandomListItem-select').simulate('click');
    expect(props.selectFandom.mock.calls.length).toEqual(2);
  });
});
