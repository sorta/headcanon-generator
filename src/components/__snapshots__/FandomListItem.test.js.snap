// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FandomListItem in managing mode renders and matches our snapshot 1`] = `
<li
  className="FandomListItem list-item control-row"
>
  <input
    checked={true}
    className="cb-fake"
    id="isAvailable_fand_f_test_1"
    name="isAvailable"
    onChange={[Function]}
    type="checkbox"
  />
  <label
    htmlFor="isAvailable_fand_f_test_1"
  />
  <input
    className="control-stretch"
    name="name"
    onChange={[Function]}
    placeholder="Fandom Name"
    type="text"
    value="Test 1"
  />
  <button
    className="FandomListItem-delete"
    onClick={[Function]}
  >
    -
  </button>
  <button
    className="FandomListItem-select"
    onClick={[Function]}
  >
    &gt;&gt;
  </button>
</li>
`;

exports[`FandomListItem renders and matches our snapshot 1`] = `
<li
  className="FandomListItem list-item control-row"
>
  <input
    checked={true}
    className="cb-fake"
    id="isAvailable_fand_f_test_1"
    name="isAvailable"
    onChange={[Function]}
    type="checkbox"
  />
  <label
    className="FandomListItem-text control-stretch"
    htmlFor="isAvailable_fand_f_test_1"
  >
    Test 1
  </label>
  <button
    className="FandomListItem-select"
    onClick={[Function]}
  >
    &gt;&gt;
  </button>
</li>
`;
